Core implementation of Professional Email upsell content

daledupreez opened the pull request:
#### Changes proposed in this Pull Request

* This PR builds on PR #56671 to flesh out the Professional Email upsell from a content and basic interaction perspective. For now, it includes the following features:
  - Content from i2 design - pcDL0W-I2-p2#comment-963
  - Mobile, tablet, and desktop layouts -- the tablet sizes are a hybrid with the buttons horizontal; please take a look at those sizes
  - Logic to skip adding email
  - Validation of mailbox fields
  - Logic to add the mailbox to the cart _(that doesn't mean it's fully working however -- more below)_
* The following items are _not_ addressed in this PR:
  - We are not wiring in the domain -- we're using a hard-coded `example.com`, which will fail validation when adding the product to the cart
  - We don't have back end support for having some of the mailbox fields be optional - that will be addressed separately

**Feedback requested**
I have not maintained 100% fidelity to the Figma design, with the biggest area of variance being padding and margins, where I've tried to aim for `em`-based values that are pretty close to the designs. I have also incorporated a lot of feedback from @saygunnyc provided via Slack (in this thread: p1633001245007300-slack-C01TJT42PMM) and in this P2 thread: pcDL0W-I2-p2#comment-1012

The other area that needs feedback is the tablet layout, where I've shifted to a columnar layout, but have kept the buttons horizontal. The biggest thing this does is reduce the prominence of the "Why Professional Email?" text -- LMK if you think this text needs to be more prominent or separated.

#### Testing instructions

* As in PR #56671, the first thing we need to do is find an existing receipt ID for one of your sites. Follow these steps to find a site slug and receipt ID:
  - For a site with purchases on it, navigate to Upgrades -> Purchases and then click on the "Billing History" tab
  - For one of the items in the list, click on the "View receipt" link
  - Look at the URL and note the last two `/`-separated pieces -- these are the site slug and receipt ID (in that order)
* Run this branch locally or via the live branch (via the link in [this comment](https://github.com/Automattic/wp-calypso/pull/56690#issuecomment-930988540))
* Manually type the following path into the browser location box: `/checkout/offer-professional-email/:receiptId/:siteSlug` with `:receiptId` and `:siteSlug` replaced by the two values above (which we use in the reverse order for this URL)
* Verify that you see the new content using a domain name of `example.com`
* Try viewing the content at different screen sizes, especially mobile, tablet, and desktop widths - verify that things look OK, are easy to navigate, and match the designs
* Try adding data to the two input fields, including empty/invalid data -- when you enter invalid data, you should see warnings
* Try refreshing the page and immediately clicking the "Add Professional Email" button -- you should see validation errors
* Try entering a valid mailbox name and password, and then click "Add Professional Email"
* Verify that you see an error stating that you don't own the domain -- you can take a look at the network traffic to verify that we're sending valid product data with the `extra`, `quantity`, and `new_quantity` data fields specified

#### Screenshots

##### Desktop

<img width="1389" alt="Screenshot 2021-09-30 at 22 19 46" src="https://user-images.githubusercontent.com/3376401/135525360-20eab5b4-9b06-4e70-b8a0-6fbfa0ad0276.png">

##### Tablet

<img width="768" alt="Screenshot 2021-09-30 at 22 20 22" src="https://user-images.githubusercontent.com/3376401/135525356-b68461a5-7112-4fcb-b475-c6cb5f4f1534.png">

##### Mobile

<img width="375" alt="Screenshot 2021-09-30 at 22 20 47" src="https://user-images.githubusercontent.com/3376401/135525353-dbf2d3a3-1685-4f31-9c7f-4e2226066492.png">
<img width="375" alt="Screenshot 2021-09-30 at 22 20 55" src="https://user-images.githubusercontent.com/3376401/135525337-f4861762-e9b8-4855-b1e8-222eda3793f5.png">

Related to #56671

daledupreez committed:
Core implementation of email upsell page, including cart update but no domain wiring

a8ci18n commented on client/my-sites/checkout/upsell-nudge/professional-email-upsell/index.jsx:
```diff
@@ -1,9 +1,225 @@
+
+	const optionalMailboxFields = [ 'alternativeEmail', 'name' ];
+
+	const formattedPrice = translate( '{{price/}} /mailbox /month (billed monthly)', {
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **17** times:
`translate( '{{price/}} /user /month billed monthly' )` *ES Score: 8*

a8ci18n commented on client/my-sites/checkout/upsell-nudge/professional-email-upsell/index.jsx:
```diff
@@ -1,9 +1,225 @@
+							{ translate( "Trusted email address that's truly yours" ) }
+						</ProfessionalEmailFeature>
+						<ProfessionalEmailFeature>
+							{ translate( 'Increase your credibility' ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **17** times:
`translate( 'Increase your brand credibility' )` *ES Score: 11*

stephanethomas commented on client/my-sites/checkout/upsell-nudge/professional-email-upsell/index.jsx:
```diff
@@ -1,9 +1,225 @@
+		comment: '{{price/}} is the formatted price, e.g. $20',
+	} );
+
+	const isEligibleForFreeTrial = true;
```
How are we going to determine if it is eligible? Do we have to call the backend for this?

daledupreez replied:
This is part of the work to wire in the correct domain - the data from the `/domains` endpoint includes information about whether a particular domain is eligible for the free trial.

stephanethomas commented on client/my-sites/checkout/upsell-nudge/professional-email-upsell/index.jsx:
```diff
@@ -1,9 +1,225 @@
 	return (
-		<div>
-			<span>Professional Email Upsell placeholder</span>
+		<div className="professional-email-upsell__main">
```
Do we really need that `professional-email-upsell__main` CSS class? It's not used anywhere.

daledupreez replied:
I will remove as part of some other outstanding items that were flagged as feedback on the design post.

daledupreez committed:
Switch back to standard colour scheme

daledupreez committed:
Remove lines between features

daledupreez committed:
Remove unnecessary class

daledupreez committed:
Left-align checks next to features

stephanethomas commented:
The code looks solid, and works fine:

<img width="623" alt="PROOF" src="https://user-images.githubusercontent.com/594356/135462193-e9cab7f3-3d5b-4707-87fe-38f947856160.png">

I'm a bit concerned by the duplication of the form logic as it will be very easy to forgot to update one of those forms whenever we have to fix one of them.

I've noticed that the placeholder doesn't really match the shape of the form, is it something we plan on addressing later?

> €3,75 /mailbox /month (billed monthly) 3 months free

Finally, the `(billed monthly)` mention seems unnecessary to me here.

saygunnyc commented:
> Finally, the (billed monthly) mention seems unnecessary to me here.

👍 Agree. As we already say /month.

daledupreez committed:
Multiple improvements from design review
 - Add "Powered by Titan" image with vertical layout and capitalised "Powered By"
 - Use the new image in the upsell page
 - Remove unnecessary `React` import
 - Remove unnecessary "(billed monthly)" text
 - Remove vertical separator between form and features
 - Let content centre across the page instead of having the separator be centred
 - Left-align the "Powered by Titan" image

daledupreez committed:
Actually commit left alignment for logo

daledupreez commented:
Thanks for the thorough review, @stephanethomas!

> I've noticed that the placeholder doesn't really match the shape of the form, is it something we plan on addressing later?

This is tricky -- the placeholder comes from the core upsell nudge. I've added the following Asana card to track that: 0-as-1201093507079510/f

daledupreez commented:
> > Finally, the (billed monthly) mention seems unnecessary to me here.
> 
> 👍 Agree. As we already say /month.

Addressed in my most recent round of commits.

saygunnyc commented:
One note based on the mobile screenshot shared above.
That would be great to reduce the font size and line height, so "Add Professional Email" fits in a single line and reads better.

<img width="579" alt="Screen Shot 2021-10-01 at 12 05 53" src="https://user-images.githubusercontent.com/78876323/135594968-a5f8de8f-fc1e-48e2-8dc5-8f64de38054f.png">

stephanethomas commented:
> That would be great to reduce the font size and line height, so "Add Professional Email" fits in a single line and reads better.

I would be wary of this kind of refinements as this only works in English, and not for other languages.

stephanethomas commented:
A heads-up that I've rebased that branch against `trunk`.

AllTerrainDeveloper reviewed (approved):
Looks good to me!

**Evidence of Testing**

Resize for mobile/desktop/tablet
![upsell-resize](https://user-images.githubusercontent.com/5689927/135980923-9a959ed7-8372-40c1-8400-c677fa0d0345.gif)

Validation errors (password is 10 chars length? I've seen it as 12 in othe forms)
![image](https://user-images.githubusercontent.com/5689927/135981153-6721bb00-14bc-4b03-8058-b176e0f03216.png)

Error stating that I don't own the domain:
![image](https://user-images.githubusercontent.com/5689927/135981602-f3426ab2-e891-4da4-bc20-9dfc8f25e997.png)

saygunnyc commented:
> I would be wary of this kind of refinements as this only works in English, and not in other languages.

@stephanethomas  Can you further explain this?
I don't understand what line height has to do with language support.

Currently, the extensive spacing between those 2 lines –"Add Professional" and "Email @example.com" both make the text hard to read and hierarchically confusing as the pricing block now has more proximity to the title than the spacing between the two lines of the header.

<img width="412" alt="Screen Shot 2021-10-05 at 13 13 57" src="https://user-images.githubusercontent.com/78876323/136004549-aa2daa0e-015c-4214-871b-6d0d54dcbae5.png">

stephanethomas commented:
Sure thing! I understood from your initial comment that you wanted to reduce the font size so that `Add Professional Email` would fit on one line. I think it's usually better to avoid optimizing the display for English given that there is no guarantee it will work in other languages, e.g. in French this string would be slightly longer (`Ajouter une adresse E-mail Professionnelle`).
You make good points about the height of the line though, and I agree it has nothing to do with the language of the user.

stephanethomas committed:
Titan: Refine font size and line height of heading

stephanethomas commented:
I've refined the font size and line height:

<img width="244" alt="screenshot" src="https://user-images.githubusercontent.com/594356/136034572-aab9671f-5084-4a87-b0ab-ff3258238629.png">

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6729332

Thank you @daledupreez for including a screenshot in the description! This is really helpful for our translators.

saygunnyc commented:
I appreciate it @stephanethomas! 🙌 One last small ask;

Would you mind adding a bit more spacing between the header and pricing for similar reasons? Sharing a quick mockup below.


![notes](https://user-images.githubusercontent.com/78876323/136042533-5d95c092-844c-4383-a54a-839c89f477f3.png)

stephanethomas commented:
> Would you mind adding a bit more spacing between the header and pricing for similar reasons? 

This will be addressed in https://github.com/Automattic/wp-calypso/pull/58240 🙂.
